### PR TITLE
4091 - Add truncated style to Breadcrumbs

### LINKED ISSUE #4091: Breadcrumb: Truncate "hidden" items, display important items
**Is your feature request related to a problem or use case? Please describe.**
A design enhancement to the IDS Breadcrumb has been requested with the following requirements:

>  The breadcrumb UI pattern is widely used and effectively gives the user an overview of where she is in a given navigation path. It also provides a simple way to navigate to previous states or pages. Today there are css classes in IDS that can be used to create breadcrumbs, but there is no component with clear examples and the current breadcrumb implementation is lacking two important features.
> - Correct truncation.
> - Possibility to see truncated parts of the breadcrumb path
> ...
> - Breadcrumbs should be truncated when they don't fit on the screen and the lowest level breadcrumb should always be visible. Today it's the first thing that disappears.
> - It should be possible to view and click parts of the breadcrumbs that has been truncated, so the user can navigate to that state. One example of this is the file system navigation bar on most computer operating systems.

After working through trying to implement these requirements into the current design, I've discovered the following pitfalls that need some design guidance:

- Our current breadcrumb design had only anticipated a handful of breadcrumb items being present at once, and never implemented "truncation" in the way this request is expecting.  Currently the line of breadcrumbs will wrap to a second line if it gets long enough.
- If we squish down the size of the breadcrumbs to a single line, and allow for expanding the size of "important" ones, we need to make design decisions on things like:
    - how many breadcrumbs should ever be visually present at once on a single line?
    - do we ever "throw away" any breadcrumbs on the line?
    - does clicking a breadcrumb automatically expand it?
    - should we have the standard "show truncated text in a tooltip" design pattern for breadcrumb items?  

**Additional context**
Originally listed as a requirement in infor-design/enterprise-ng#700, per @gushanson's request

**Examples**
Screen Shot 2020-07-07 at 4 41 24 PM
Screen Shot 2020-07-07 at 4 41 15 PM

### PR DESCRIPTION
**Pull request by EdwardCoyle**
**Explain the _details_ for making this change. What existing problem does the pull request solve?**
This PR:
- Adds a truncated style to the Breadcrumb component, per request from @gushanson
- Sets this new style to the default (old wrapping style is still optionally available / used as a fallback in IE11)
- Makes various visual tweaks to all Breadcrumb variants (changes in padding/spacing/etc)
- Adds a missing API endpoint (`get disabled() on individual breadcrumbs`)
- Adds/updates tests for the new truncation feature

**Related github/jira issue (required)**:
Closes #4091

**Steps necessary to review your pull request (required)**:
Pull this branch, build, run the demoapp. Then test all scenarios below.  On all scenarios, also all combinations of RTL, Uplift, Personalization (all should be working).

**NOTE:** The resize detection was created using a `ResizeObserver`.  This primitive doesn't exist in IE11, so if you test this feature in IE11, the breadcrumb list does not truncate.  However, it should gracefully degrade back to the original, non-truncated style in this case, so it's worth testing.

_Test basic samples:_
- Open http://localhost:4000/components/breadcrumb/example-index.html
- When this page loads at a large form-factor, no truncation should occur.
- Also open http://localhost:4000/components/breadcrumb/example-from-settings.html
- When this page loads at a large form-factor, no truncation should occur.
- Clicking Destroy/Invoke should work without causing any JS console errors.

_Test truncation:_
- Open http://localhost:4000/components/breadcrumb/example-add-remove.html
- Click the "Add" button to add enough breadcrumbs to the list to cause truncation.
- Click the "..." button that appears to the left of the breadcrumbs to open the overflow menu.  When it opens, it should display all list items that are cut off and/or hidden from the list.
- Click any of the items in the overflow menu.  When clicking, a toast should display with the correct Breadcrumb name (this signals that the Breadcrumb item's callback method is properly being triggered from the overflow menu).
- Try adding/removing breadcrumbs, adding disabled breadcrumbs, and using the `Tab`/`Shift`+`Tab` keys to navigate across the Breadcrumb list. Any items that are cut off/hidden/disabled should not become focused.  The Popupmenu should open using its normal keyboard handling.

_Test header styles:_
Try the following pages to test the standard Header breadcrumb style:
- http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs.html
- http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs.html?theme=uplift
http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs.html?theme=uplift&variant=dark&colors=7928E1

Try the following pages to test the alternate Header breadcrumb style:
- http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs-alternate.html
- http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs-alternate.html?theme=uplift
http://localhost:4000/components/breadcrumb/example-navigation-breadcrumbs-alternate.html?theme=uplift&variant=dark&colors=7928E1

Try various combinations of themes/personalization.

_Test where `truncate: false` is set:_
- Open http://localhost:4000/components/breadcrumb/example-no-truncate.html
- Observe this breadcrumb list will always wrap and never truncate.

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
- [x] A note to the change log.

### CODE REVIEW

File: docs/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 ### v4.32.0 Features
 
+- `[Breadcrumb]` Add truncated style and made it the default for all Breadcrumb lists. `EPC` ([#4091](https://github.com/infor-design/enterprise/issues/4091))
```
Comment (tmcconechy) [inline docs/CHANGELOG.md:7]:
```suggestion
- `[Breadcrumb]` Add truncated style and made it the default for all Breadcrumb lists. ([#4091](https://github.com/infor-design/enterprise/issues/4091))
```

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,7 @@
...
 - `[Datagrid]` Add a new `RowNumber` formatter that will show a row number column that remains the same no matter how the grid is sorted. ([#1904](https://github.com/infor-design/enterprise/issues/1904))
 - `[Datepicker]` Added the ability to use the range selection in date picker when using the UmAlQura Calendar (RTL). `TJM` ([#4227](https://github.com/infor-design/enterprise/issues/4227))
```
Comment (tmcconechy) [inline docs/CHANGELOG.md:9]:
```suggestion
- `[Datepicker]` Added the ability to use the range selection in date picker when using the UmAlQura Calendar (RTL). ([#4227](https://github.com/infor-design/enterprise/issues/4227))
```

File: src/components/breadcrumb/breadcrumb.js [hunk 3]
```diff
@@ -104,6 +107,23 @@ BreadcrumbItem.prototype = {
     return li;
   },
 
+  /**
+   * Set the anchor's tabIndex based on its current overflow state, if applicable
+   * @returns {void}
+   */
+  checkFocus() {
```
Comment (tmcconechy) [inline src/components/breadcrumb/breadcrumb.js:114]: Is this private?

File: src/components/breadcrumb/breadcrumb.js [hunk 4]
```diff
@@ -104,6 +107,23 @@ BreadcrumbItem.prototype = {
...
+    const a = this.a;
+    a.tabIndex = (this.overflowed || this.disabled) ? -1 : 0;
+    a.setAttribute('tabindex', a.tabIndex);
+  },
+
+  /**
+   * @returns {HTMLAnchorElement} reference to this breadcrumb item's anchor tag
+   */
+  get a() {
```
Comment (tmcconechy) [inline src/components/breadcrumb/breadcrumb.js:123]: And this?

Comment (EdwardCoyle) [reply src/components/breadcrumb/breadcrumb.js:123]: I actually used this one in the tests, I think it's OK to expose if you've got access to the breadcrumb API

File: src/components/breadcrumb/breadcrumb.js (hunk 3 above)
Comment (EdwardCoyle) [reply src/components/breadcrumb/breadcrumb.js:114]: This one probably should be

File: src/components/breadcrumb/breadcrumb.js [hunk 5]
```diff
@@ -512,6 +625,31 @@ Breadcrumb.prototype = {
...
+
+    // Picking an item from the overflow menu should cause the original breadcrumb item's operation to occur.
+    // This will either trigger the item's callback, or simply follow its `href` attribute.
+    if (this.overflowBtn) {
+      $(this.overflowBtn).on(`selected.${COMPONENT_NAME}`, (e, ...args) => {
+        // First argument is the clicked item from the `popupmenu.selected` event
+        const liItem = args[0];
+        const index = liItem[0].getAttribute('data-breadcrumb-index');
+        const breadcrumbAPI = this.overflowed[Number(index)];
+        $(breadcrumbAPI.a).trigger('click');
```
[the remark this answers is not in the document]
Comment (pwpatton) [reply src/components/breadcrumb/breadcrumb.js:649]: Nevermind, I debugged and found it all goes through the callback :)

**Comment by pwpatton** [comment]
This looks great!

**Comment by deep7102** [comment]
This looks so good! found few also as:

need to fix this destroy example page, because destroy not working here http://localhost:4000/components/breadcrumb/example-destroy-invoke.html but working here http://localhost:4000/components/breadcrumb/example-from-settings.html

safari - cannot get to breadcrumbs by keyboard
http://localhost:4000/components/breadcrumb/example-callback.html

ie11 - can not load breadcrumbs
http://localhost:4000/components/breadcrumb/example-callback.html

ie11 and edge - can not get to first breadcrumb (Home) by keyboard
http://localhost:4000/components/breadcrumb/example-index.html

**Comment by EdwardCoyle** [comment]
@deep7102 I think I addressed the issues with tabIndex.  As far as Safari goes, make sure the setting for respecting all tabIndexes is active in your Safari Preferences.  I wasn't seeing that issue until I tried turning it off:

Screen Shot 2020-08-27 at 11 53 50 AM

Regarding the "remaining" breadcrumbs on the page here: http://localhost:4000/components/breadcrumb/example-destroy-invoke.html -- this is actually a built in feature for backwards compatibility.  If you generate Breadcrumbs via HTML markup instead of by doing it with a dataset, the ones generated from HTML hang around until explicitly removed.  On the page where all the breadcrumbs disappear, all of them are generated from data instead.

**Pull request merged**